### ISSUE TITLE
why do you compute grad and do SGD mannualy ?

### ISSUE THREAD

**Issue by schinger** [maple-research-lab/AdCo#5]
In train.py Line 60-67, why do you compute grad and do SGD mannualy ? can we leverage pytorch auto computed grad, multipy by -1 and then SGD by optimizer for negative bank? this will be simple.

and you use momentum encode(stop SGD) for k, you don't mention this in paper, does this important for performance?

Thanks!

**Comment by wang3702** [comment]
Thanks for your interest!
1) Yes, you can do that, but we found our manual updates will be faster. The optimizer version will be slower, that's why we adopt this manual updates. Simply adding a -1 for the loss can work, but please pay attention to that the temperature is different from the temperature for the contrastive loss for the encoder.
2) We used the MoCo as comparison, thus we directly followed the design of momentum encoder in moco. In MoCo, only query encoder is updated, the same as in AdCo.
